tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,176 @@
+/*!
+Packages using the Go programming language may have upstream tar archives that
+include only the source code of the project, but not the source code of any
+dependencies. The Go programming language promotes the use of "modules" for
+dependencies and projects adopting modules will provide go.mod and go.sum
+files.
+
+This module extends the functionality of `packages.metadata.build-package.external-files`
+and provides the ability to retrieve and validate dependencies
+declared using Go modules given a tar archive containing a go.mod and go.sum.
+
+The location where dependencies are retrieved from are controlled by the
+standard environment variables employed by the Go tool: GOPROXY, GOSUMDB, and
+GOPRIVATE.
+
+ */
+
+pub(crate) mod error;
+use error::Result;
+
+use super::manifest;
+use duct::cmd;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::env;
+use std::path::{Path, PathBuf};
+use std::process::Output;
+
+pub(crate) struct GoMod;
+
+impl GoMod {
+    pub(crate) fn vendor(
+        root_dir: &Path,
+        package_dir: &Path,
+        gomod: &manifest::ExternalFile,
+    ) -> Result<()> {
+        let url_file_name = extract_file_name(&gomod.url)?;
+        let path_arg = &gomod.path.as_ref().unwrap_or(&url_file_name);
+        ensure!(path_arg.components().count() == 1, error::InputFileSnafu);
+
+        let path = package_dir.join(path_arg);
+        ensure!(path.is_file(), error::InputFileBadSnafu { path });
+        let mod_dir = gomod.bundle_path.as_ref().context(error::ModDirSnafu)?;
+        let output_path_arg = gomod
+            .bundle_output_path
+            .as_ref()
+            .context(error::OutputDirSnafu)?;
+        let output_path = package_dir.join(output_path_arg);
```

Comment:
Somewhere, maybe right here, we need to emit `rerun-if-changed` for the output bundle archive.

Reply:
👍🏼 Done!